### PR TITLE
Refactor in process plasma store interface

### PR DESCRIPTION
**Pull request by MissiontoMars**
## Why are these changes needed?
This is an early prototype implementation of in-process plasma store access.
1. Abstract the `PlasmaClientInterface` interface for plasma client, and derived two subclasses, the one is `PlasmaClient` for in-process access, the other `RemotePlasmaClient` for cross-process access from CoreWorker.
2. Extract some common logic from `PlasmsStore::ProcessMessage`, such as `PlasmsStore::CreateObject`, which is called by both in-process and cross-process.
3. When making an in-process call, use `PlasmaStore::ExecuteInStoreThread` to ensure the store interface executes within that store internal io_context_.
4. This early PR shows the prototype of creating an object, which may be the most complex operation. Others should be simpler.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/plasma/client.h [hunk 1]
```diff
@@ -210,23 +197,106 @@ class PlasmaClient {
...
+  virtual Status Evict(int64_t num_bytes, int64_t &num_bytes_evicted) = 0;
+
+  /// Get the current debug string from the plasma store server.
+  ///
+  /// \return The debug string.
+  virtual std::string DebugString() = 0;
+};
+
+class Impl;
+class PlasmaClient : public PlasmaClientInterface {
```
Comment (scv119) [inline src/ray/object_manager/plasma/client.h:209]: cc @rkooo567 for the new unified plasma client interface. 
like it a lot!

**Comment by scv119** [comment]
At a high level I really like the direction of unifying the interface of accessing plasma store from a different process vs in the same process.

**Comment by stephanie-wang** [comment]
Nice, this also looks like it will make testing classes that use a plasma client much nicer!

One thought, would the PlasmaClient be usable if we had it inherit from RemotePlasmaClient and override individual methods incrementally? Then we could start doing e2e testing right away.

**Comment by MissiontoMars** [comment]
> Nice, this also looks like it will make testing classes that use a plasma client much nicer!
> 
> One thought, would the PlasmaClient be usable if we had it inherit from RemotePlasmaClient and override individual methods incrementally? Then we could start doing e2e testing right away.

Yeah, more unit tests will be added after this refactor.
A little confusing about the class inheritance. `PlasmaClient` **is not a** `RemotePlasmaClient`,see: https://stackoverflow.com/questions/2218937/has-a-is-a-terminology-in-object-oriented-language
In this prototype, the PlasmaClient and the RemotePlasmaClient share common code using the legacy class `Impl`.

**Comment by stephanie-wang** [comment]
> > Nice, this also looks like it will make testing classes that use a plasma client much nicer!
> > One thought, would the PlasmaClient be usable if we had it inherit from RemotePlasmaClient and override individual methods incrementally? Then we could start doing e2e testing right away.
> 
> Yeah, more unit tests will be added after this refactor.
> A little confusing about the class inheritance. `PlasmaClient` **is not a** `RemotePlasmaClient`,see: https://stackoverflow.com/questions/2218937/has-a-is-a-terminology-in-object-oriented-language
> In this prototype, the PlasmaClient and the RemotePlasmaClient share common code using the legacy class `Impl`.

Ah got it, thanks! I was just wondering about this because the other methods for PlasmaClient weren't implemented yet, but your approach makes sense.

**Comment by rkooo567** [comment]
@MissiontoMars lmk when you addressed the initial feedback. I will review the PR after

**Comment by rkooo567** [comment]
Is this still WIP?

**Comment by MissiontoMars** [comment]
> Is this still WIP?

Development comes to an end, we can review this pr now. I'll remove the 'WIP'.

**Comment by rkooo567** [comment]
So, this is a really nice change, but I think we shouldn't merge this as it is. It is too large PR, meaning it is easy to miss important changes and more likely to break the master.

Is it possible to split the PRs out into multiple of them?

**Comment by MissiontoMars** [comment]
> 

That make sense. Will contribute several PRs.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**